DM-24260: Create Gen3 versions of ap_verify datasets

kfindeisen opened the pull request:
This PR adds a script that converts all Gen 2 content in the dataset to Gen 3, along with config files the script requires. Since all `ap_verify` datasets are created from this repository, this adds a copy of the script and stub configs to the new dataset. The `script/` directory has already been used by individual datasets to house maintenance code, although this is its first use in `dataset_template`.

For background, Gen 2 `ap_verify` datasets store all their data (except coadded templates) in *uningested* form, as separate directories of raw files, calib files, etc. `ap_verify` creates a new repository and ingests the data as the first step of its processing. From various discussions with AP members, it was decided that, in Gen 3, datasets would provide a pre-ingested repository of everything except raws, and that `ap_verify` would create a new repository from the dataset and ingest the raws there (see [DM-21915](https://jira.lsstcorp.org/browse/DM-21915) for a summary).

In both Gen 2 and Gen 3, the downloaded dataset should not be altered by any Butler operation, to keep us from accidentally committing processed data and to provide a "clean slate" every time `ap_verify` is run. In Gen 3, this means providing an export file that can be used to copy the pre-ingested repository.

To minimize download size, the Gen 3 repository uses relative symbolic links to the data that was already stored for Gen 2. This is not possible for refcats (which were previously stored in `.tar.gz` files) or camera/transmission/defect data (which are imported from outside the dataset during conversion); these are copied instead.

kfindeisen committed:
Create Gen 3 conversion script.

This script can be used to maintain ap_verify datasets that have both
Gen 2 and Gen 3 data, keeping them in sync. It can also perform the
final migration once we no longer need Gen 2 datasets.

kfindeisen committed:
Add template config files for Gen2->Gen3 conversion.

These files should work for most datasets, once the list of refcats
is updated.

TallJimbo commented on config/convertRepo_calibs.py:
```diff
@@ -0,0 +1,4 @@
+# Config overrides for convert_gen2_repo_to_gen3.py
+
+config.datasetIgnorePatterns.append("raw")
+config.datasetIgnorePatterns = ["raw", "*Coadd_skyMap", "ref_cat", "defects"]
```
Looks like the first of these two lines has no effect.

TallJimbo commented on scripts/add_gen3_repo.py:
```diff
@@ -0,0 +1,149 @@
+                          help="The name of the dataset as recognized by ap_verify.py.")
+        self.add_argument("--drop-gen2", action="store_true",
+                          help="Create a standalone Gen 3 repo instead of sharing files with Gen 2. "
+                               "Intended for use only once ap_verify no longer supports Gen 2.")
```
Why is this a subclass of `ArgumentParser` rather than a factory function?

kfindeisen replied:
Because `argparse` is an object-oriented CLI library, and most such libraries are designed on the assumption that they will be subclassed by the application. I haven't seen any Python documentation specifically saying to do it that way, but it's basically my default.

TallJimbo commented on scripts/add_gen3_repo.py:
```diff
@@ -0,0 +1,149 @@
+                                  # skymap detected automatically; do nothing
+                                  calibs=gen2_calib_repo,
+                                  config=config,
+                                  transferMode=mode)
```
Did you consider using `ConvertRepoTask` directly instead of the command-line-driver wrapper function here?  I don't want to get in the way of you getting things done by asking you to change things unnecessarily, but I'd like the `Task` to be the interface usually used in Python code (as is usual for `Tasks`), so if there's anything the script does that it can't, I'd like to know about it.

(We do have test code that calls the command-line-script wrapper function, but that's just to test as much of the command-line script as we can.)

kfindeisen replied:
Yes, I did consider it, but then I would have to duplicate about 30 lines of code (https://github.com/lsst/obs_base/blob/master/python/lsst/obs/base/script/convertGen2RepoToGen3.py#L110-L138), manually setting up separate skymap configs (though I don't understand why that's necessary) and a Butler. The code would then need to be kept in sync with any changes to `obs_base`, which is difficult given that an isolated script like this is susceptible to bit rot.

This approach seems more maintainable, because it delegates as much logic as possible to well-maintained code.

TallJimbo commented on scripts/add_gen3_repo.py:
```diff
@@ -0,0 +1,149 @@
+       datasets, or an independent copy.
+    mode : `str`
+       The config file (in the dataset config directory) with a configuration
+       for `~lsst.obs.base.gen2to3.ConvertRepoTask`
```
Looks like this doc belongs to `config_file`, not `mode`.

TallJimbo commented on scripts/add_gen3_repo.py:
```diff
@@ -0,0 +1,149 @@
+    try:
+        return classes[instrument]
+    except KeyError:
+        raise ValueError(f"Unsupported instrument {instrument}; consider registering its Instrument class.")
```
This might be a slightly misleading error message, if it's interpreted as a request to register a Gen3 `Instrument` class with the the registry (which of course won't modify this dict).

timj replied:
Yes, it means adding it to the list here and not adding to registry. I've been worrying about short name to class name mapping for a long time. We've mitigated a bit now that registry can do that look up for you but we've never solved this problem in the general case where you are seeding a repo. If we are going to start adding dicts like this all over the code we should add them to the obs_base code that does this already by registry lookup (falling back to a dict if not registered). If we are doing that we could also remove the need for specifying an instrument class in 2to3 conversion by including a dict that maps gen2 mapper to instrument class.

kfindeisen replied:
I'll reword the message to avoid using the word "register".

@timj, while I'd love to see a centralized/automated solution, note that in this case it's unlikely that we will have `ap_verify` datasets for instruments other than HSC, DECam, and LSSTCam (I'll deal with that case on DM-24844), at least not before we stop supporting Gen 2 in `ap_verify`. So a dict here is a bit more maintainable than it would be in general-purpose code, where it would need to be kept up to date with the full list of known obs packages.

So I suggest leaving it as-is, and removing/re-iplementing once a centralized lookup is available.

TallJimbo reviewed (approved):
Looks good!  My only substantial comment is more of a question-for-my-benefit than a concern; besides that, just a few minor cleanup things.

kfindeisen committed:
Document Gen 3 repository format.

This format defines standard collection names that can be used by
ap_verify to set up a valid Butler. Most names are defaults
provided by the Gen 2 -> Gen 3 converter.
